vpr/src/timing/PreClusterTimingManager.h

```diff
@@ -0,0 +1,99 @@
+/**
+ * @file
+ * @author  Alex Singer
+ * @date    April 2025
+ * @brief   Manager class for pre-cluster (primitive-level) timing analysis.
+ */
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include "vpr_types.h"
+#include "vtr_assert.h"
+
+// Forward declarations.
+class AtomLookup;
+class AtomNetlist;
+class PreClusterDelayCalculator;
+class Prepacker;
+class SetupTimingInfo;
+
+/**
+ * @brief Pre-cluster timing manager class.
+ *
+ * This class encapsulates the timing computations used prior to clustering.
+ * This maintains all of the state necessary to perform these timing computations.
+ */
+class PreClusterTimingManager {
+  public:
+    /**
+     * @brief Constructor for the manager class. Builds the timing graph.
+     *
+     *  @param timing_driven
+     *          Whether this class should compute timing information or not. This
+     *          may seem counter-intuitive, but this class still needs to exist
+     *          even if timing is turned off. This will not initialize anything
+     *          and set the valid flag to false if we are not timing driven.
+     *  @param atom_netlist
+     *          The primitive netlist to perform timing analysis over.
+     *  @param atom_lookup
+     *          A lookup between the primitives and their timing nodes.
+     *  @param prepacker
+     *          The prepacker object used to prepack primitives into molecules.
+     *  @param timing_update_type
+     *          The type of timing update this class should perform.
+     *  @param arch
+     *          The architecture.
+     *  @param routing_arch
+     *          The routing architecture.
+     *  @param analysis opts
+     *          Options for the timing analysis in VPR.
+     */
+    PreClusterTimingManager(bool timing_driven,
+                            const AtomNetlist& atom_netlist,
+                            const AtomLookup& atom_lookup,
+                            const Prepacker& prepacker,
+                            e_timing_update_type timing_update_type,
+                            const t_arch& arch,
+                            const t_det_routing_arch& routing_arch,
+                            const std::string& device_layout,
+                            const t_analysis_opts& analysis_opts);
+
+    /**
+     * @brief Calculates the setup criticality of the given primitive block.
```

Comment:
Define what the setup criticality of a block is.  Currently defined as the maximum criticality over the block inputs.